tools/codegen/gen_backend_stubs.py

```diff
@@ -100,44 +110,66 @@ def make_file_manager(install_dir: str) -> FileManager:
-        'generated_comment': generated_comment,
-        'cpp_namespace': cpp_namespace,
-        'dispatch_xla_declarations': list(concatMap(dest.compute_native_function_declaration, external_backend_functions)),
-    })
-
-    fm.write('aten_xla_type_default.h', lambda: {
-        'generated_comment': generated_comment,
-        'cpp_namespace': cpp_namespace,
-        'dispatch_aten_fallback_declarations': list(concatMap(
-            dest.GenExternalAtenFallback(Target.NAMESPACED_DECLARATION), external_backend_functions
-        )),
-    })
-
-    fm.write('aten_xla_type_default.cpp', lambda: {
-        'generated_comment': generated_comment,
-        'cpp_namespace': cpp_namespace,
-        # TODO: after cpu fallbacks are moved to a boxed kernel,
-        # merge registrations / definitions into RegisterDispatchKey
-        'dispatch_aten_fallback_definitions': list(concatMap(
-            dest.GenExternalAtenFallback(Target.NAMESPACED_DEFINITION), external_backend_functions
-        )),
-        'dispatch_registrations': list(concatMap(
-            dest.GenExternalAtenFallback(Target.REGISTRATION), [e for e in external_backend_functions if not e.is_autograd_kernel]
-        )),
-        'dispatch_autograd_registrations': list(concatMap(
-            dest.GenExternalAtenFallback(Target.REGISTRATION), [e for e in external_backend_functions if e.is_autograd_kernel]
-        )),
-    })
+    # TODO: handle cases when yaml contains zero ops properly in a later PR.
+    if backend_key is not None and autograd_key is not None:
+        backend_dispatch_key: DispatchKey = backend_key
+        autograd_dispatch_key: DispatchKey = autograd_key
+        generated_comment = 'Autogenerated file by gen_backend_stubs.py. Do not edit directly!'
+        fm.write('aten_xla_type.h', lambda: {
+            'generated_comment': generated_comment,
+            'cpp_namespace': cpp_namespace,
+            # Convert to a set first to remove duplicate kernel names.
+            # Backends are allowed to repeat kernel names; only generate the declaration once!
+            'dispatch_xla_declarations': list(set(concatMap(
+                lambda f: dest.compute_native_function_declaration(f, backend_indices[backend_dispatch_key]),
+                grouped_native_functions
+            ))) + list(set(concatMap(
+                lambda f: dest.compute_native_function_declaration(f, backend_indices[autograd_dispatch_key]),
+                grouped_native_functions
+            ))),
+        })
+
+        fm.write('aten_xla_type_default.h', lambda: {
+            'generated_comment': generated_comment,
+            'cpp_namespace': cpp_namespace,
+            'dispatch_aten_fallback_declarations': list(concatMap(
+                dest.GenExternalAtenFallback(Target.NAMESPACED_DECLARATION, backend_indices[backend_dispatch_key]),
+                [g for g in grouped_native_functions if not backend_indices[autograd_dispatch_key].has_backend(g)]
```

Comment:
Here's an annoying thing about GenExternalAtenFallbacks: It doesn't "just work" when you call it with multiple dispatch keys and concat the results together, since the whole point is that it generates CPU fallback kernels for everything, include operators that don't have a kernel for the specified backend. So calling it twice, once with the XLA key and one with AutogradXLA, results in a bunch of duplicates. And it also does the wrong thing for AutogradXLA, since we don't want to register CPU fallbacks to the AutogradXLA key, just XLA. I think it'll be a little better once I remove all of the non-fallback related logic from GenExternalAtenFallback.

Reply:
Splitting the codegen into two separate pieces to avoid duplication sounds like the right call.